### ISSUE TITLE
Segfault with caller()

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5469]
Migrated from [rt.perl.org#9367](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=9367) (status was 'resolved')

Searchable as RT9367$

**Comment by p5pRT** [comment]
### From scott@gossamer-threads.com

#### Created by sbeck@athena.gossamer-threads.com

This is a bug report for perl from sbeck@&#8203;athena\.gossamer\-threads\.com\,
generated with the help of perlbug 1\.33 running under perl v5\.6\.1\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
This bug started appearing in perl5\.6\.1\. Perl segfaults when you delete a named
subroutine from the symbol table\, and then call it with a reference to it\. Here
is an example with the latest perl I have&#8203;:

&nbsp;    perl5\.7\.3 \-e 'sub foo \{my @&#8203;f = caller\(0\)\} my $code = delete $&#8203;::\{foo\}; $code\->\(\);'

This causes a Segmentation fault\. This bug only applies to calling caller\(\) in
list context\, as this does not segfault&#8203;:

&nbsp;    perl5\.7\.3 \-e 'sub foo \{my $f = caller\(0\)\} my $code = delete $&#8203;::\{foo\}; $code\->\(\);'

Furthermore this only applies to calling the subroutine by a reference to it\,
these do not segfault&#8203;:

&nbsp;    perl5\.7\.3 \-e 'sub foo \{my $f = caller\(0\)\} my $code = delete $&#8203;::\{foo\}; foo\(\);'
&nbsp;    perl5\.7\.3 \-e 'sub foo \{my @&#8203;f = caller\(0\)\} my $code = delete $&#8203;::\{foo\}; foo\(\);'


Here is a bt from gdb&#8203;:

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
~$ gdb perl5\.7\.3
GNU gdb 5\.0mdk\-11mdk Linux\-Mandrake 8\.0
Copyright 2001 Free Software Foundation\, Inc\.
GDB is free software\, covered by the GNU General Public License\, and you are
welcome to change it and/or distribute copies of it under certain conditions\.
Type "show copying" to see the conditions\.
There is absolutely no warranty for GDB\.  Type "show warranty" for details\.
This GDB was configured as "i386\-mandrake\-linux"\.\.\.
\(gdb\) run \-e 'sub foo \{my @&#8203;f = caller\(0\)\} my $code = delete $&#8203;::\{foo\}; $code\->\(\);'
Starting program&#8203;: /usr/bin/perl5\.7\.3 \-e 'sub foo \{my @&#8203;f = caller\(0\)\} my $code = delete $&#8203;::\{foo\}; $code\->\(\);'
\[New Thread 1024 \(LWP 11334\)\]

Program received signal SIGSEGV\, Segmentation fault\.
\[Switching to Thread 1024 \(LWP 11334\)\]
0x080698a7 in Perl\_gv\_efullname3 \(\)
\(gdb\) bt
\#0  0x080698a7 in Perl\_gv\_efullname3 \(\)
\#1  0x080daf25 in Perl\_pp\_caller \(\)
\#2  0x080af710 in Perl\_runops\_standard \(\)
\#3  0x08061ff4 in S\_run\_body \(\)
\#4  0x08061d06 in perl\_run \(\)
\#5  0x0805f089 in main \(\)
\#6  0x400900de in \_\_libc\_start\_main \(\) from /lib/libc\.so\.6
\(gdb\) 
\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<

Here is an strace&#8203;:

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
~$ strace perl5\.7\.3 \-e 'sub foo \{my @&#8203;f = caller\(0\)\} my $code = delete $&#8203;::\{foo\}; $code\->\(\);'
execve\("/usr/bin/perl5\.7\.3"\, \["perl5\.7\.3"\, "\-e"\, "sub foo \{my @&#8203;f = caller\(0\)\} my $code = delete $&#8203;::\{foo\}; $code\->\(\);"\]\, \[/\* 40 vars \*/\]\) = 0
uname\(\{sys="Linux"\, node="jagerman"\, \.\.\.\}\) = 0
brk\(0\)                                  = 0x8137740
open\("/etc/ld\.so\.preload"\, O\_RDONLY\)    = \-1 ENOENT \(No such file or directory\)
open\("/etc/ld\.so\.cache"\, O\_RDONLY\)      = 3
fstat64\(3\, \{st\_mode=S\_IFREG|0644\, st\_size=43780\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 43780\, PROT\_READ\, MAP\_PRIVATE\, 3\, 0\) = 0x40016000
close\(3\)                                = 0
open\("/lib/libnsl\.so\.1"\, O\_RDONLY\)      = 3
read\(3\, "\\177ELF\\1\\1\\1\\0\\0\\0\\0\\0\\0\\0\\0\\0\\3\\0\\3\\0\\1\\0\\0\\0 ?\\0\\000"\.\.\.\, 1024\) = 1024
fstat64\(3\, \{st\_mode=S\_IFREG|0755\, st\_size=78208\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 4096\, PROT\_READ|PROT\_WRITE\, MAP\_PRIVATE|MAP\_ANONYMOUS\, \-1\, 0\) = 0x40021000
old\_mmap\(NULL\, 89632\, PROT\_READ|PROT\_EXEC\, MAP\_PRIVATE\, 3\, 0\) = 0x40022000
mprotect\(0x40035000\, 11808\, PROT\_NONE\)  = 0
old\_mmap\(0x40035000\, 4096\, PROT\_READ|PROT\_WRITE\, MAP\_PRIVATE|MAP\_FIXED\, 3\, 0x12000\) = 0x40035000
old\_mmap\(0x40036000\, 7712\, PROT\_READ|PROT\_WRITE\, MAP\_PRIVATE|MAP\_FIXED|MAP\_ANONYMOUS\, \-1\, 0\) = 0x40036000
close\(3\)                                = 0
open\("/lib/libdl\.so\.2"\, O\_RDONLY\)       = 3
read\(3\, "\\177ELF\\1\\1\\1\\0\\0\\0\\0\\0\\0\\0\\0\\0\\3\\0\\3\\0\\1\\0\\0\\0\\220\\36"\.\.\.\, 1024\) = 1024
fstat64\(3\, \{st\_mode=S\_IFREG|0755\, st\_size=9900\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 12832\, PROT\_READ|PROT\_EXEC\, MAP\_PRIVATE\, 3\, 0\) = 0x40038000
mprotect\(0x4003b000\, 544\, PROT\_NONE\)    = 0
old\_mmap\(0x4003b000\, 4096\, PROT\_READ|PROT\_WRITE\, MAP\_PRIVATE|MAP\_FIXED\, 3\, 0x2000\) = 0x4003b000
close\(3\)                                = 0
open\("/lib/libm\.so\.6"\, O\_RDONLY\)        = 3
read\(3\, "\\177ELF\\1\\1\\1\\0\\0\\0\\0\\0\\0\\0\\0\\0\\3\\0\\3\\0\\1\\0\\0\\0\\340I\\0"\.\.\.\, 1024\) = 1024
fstat64\(3\, \{st\_mode=S\_IFREG|0755\, st\_size=129908\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 131448\, PROT\_READ|PROT\_EXEC\, MAP\_PRIVATE\, 3\, 0\) = 0x4003c000
mprotect\(0x4005b000\, 4472\, PROT\_NONE\)   = 0
old\_mmap\(0x4005b000\, 8192\, PROT\_READ|PROT\_WRITE\, MAP\_PRIVATE|MAP\_FIXED\, 3\, 0x1e000\) = 0x4005b000
close\(3\)                                = 0
open\("/lib/libpthread\.so\.0"\, O\_RDONLY\)  = 3
read\(3\, "\\177ELF\\1\\1\\1\\0\\0\\0\\0\\0\\0\\0\\0\\0\\3\\0\\3\\0\\1\\0\\0\\0\\240O\\0"\.\.\.\, 1024\) = 1024
fstat64\(3\, \{st\_mode=S\_IFREG|0755\, st\_size=517867\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 90396\, PROT\_READ|PROT\_EXEC\, MAP\_PRIVATE\, 3\, 0\) = 0x4005d000
mprotect\(0x4006c000\, 28956\, PROT\_NONE\)  = 0
old\_mmap\(0x4006c000\, 28672\, PROT\_READ|PROT\_WRITE\, MAP\_PRIVATE|MAP\_FIXED\, 3\, 0xe000\) = 0x4006c000
old\_mmap\(0x40073000\, 284\, PROT\_READ|PROT\_WRITE\, MAP\_PRIVATE|MAP\_FIXED|MAP\_ANONYMOUS\, \-1\, 0\) = 0x40073000
close\(3\)                                = 0
open\("/lib/libc\.so\.6"\, O\_RDONLY\)        = 3
read\(3\, "\\177ELF\\1\\1\\1\\0\\0\\0\\0\\0\\0\\0\\0\\0\\3\\0\\3\\0\\1\\0\\0\\0\\0\\302\\1"\.\.\.\, 1024\) = 1024
fstat64\(3\, \{st\_mode=S\_IFREG|0755\, st\_size=1216268\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 1231496\, PROT\_READ|PROT\_EXEC\, MAP\_PRIVATE\, 3\, 0\) = 0x40074000
mprotect\(0x40198000\, 35464\, PROT\_NONE\)  = 0
old\_mmap\(0x40198000\, 20480\, PROT\_READ|PROT\_WRITE\, MAP\_PRIVATE|MAP\_FIXED\, 3\, 0x123000\) = 0x40198000
old\_mmap\(0x4019d000\, 14984\, PROT\_READ|PROT\_WRITE\, MAP\_PRIVATE|MAP\_FIXED|MAP\_ANONYMOUS\, \-1\, 0\) = 0x4019d000
close\(3\)                                = 0
open\("/lib/libcrypt\.so\.1"\, O\_RDONLY\)    = 3
read\(3\, "\\177ELF\\1\\1\\1\\0\\0\\0\\0\\0\\0\\0\\0\\0\\3\\0\\3\\0\\1\\0\\0\\0\\0\\20\\0"\.\.\.\, 1024\) = 1024
fstat64\(3\, \{st\_mode=S\_IFREG|0755\, st\_size=21812\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 184604\, PROT\_READ|PROT\_EXEC\, MAP\_PRIVATE\, 3\, 0\) = 0x401a1000
mprotect\(0x401a6000\, 164124\, PROT\_NONE\) = 0
old\_mmap\(0x401a6000\, 8192\, PROT\_READ|PROT\_WRITE\, MAP\_PRIVATE|MAP\_FIXED\, 3\, 0x4000\) = 0x401a6000
old\_mmap\(0x401a8000\, 155932\, PROT\_READ|PROT\_WRITE\, MAP\_PRIVATE|MAP\_FIXED|MAP\_ANONYMOUS\, \-1\, 0\) = 0x401a8000
close\(3\)                                = 0
open\("/lib/libutil\.so\.1"\, O\_RDONLY\)     = 3
read\(3\, "\\177ELF\\1\\1\\1\\0\\0\\0\\0\\0\\0\\0\\0\\0\\3\\0\\3\\0\\1\\0\\0\\0P\\16\\0\\000"\.\.\.\, 1024\) = 1024
fstat64\(3\, \{st\_mode=S\_IFREG|0755\, st\_size=8284\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 11188\, PROT\_READ|PROT\_EXEC\, MAP\_PRIVATE\, 3\, 0\) = 0x401cf000
mprotect\(0x401d1000\, 2996\, PROT\_NONE\)   = 0
old\_mmap\(0x401d1000\, 4096\, PROT\_READ|PROT\_WRITE\, MAP\_PRIVATE|MAP\_FIXED\, 3\, 0x1000\) = 0x401d1000
close\(3\)                                = 0
open\("/lib/libc\.so\.6"\, O\_RDONLY\)        = 3
read\(3\, "\\177ELF\\1\\1\\1\\0\\0\\0\\0\\0\\0\\0\\0\\0\\3\\0\\3\\0\\1\\0\\0\\0\\0\\302\\1"\.\.\.\, 1024\) = 1024
fstat64\(3\, \{st\_mode=S\_IFREG|0755\, st\_size=1216268\, \.\.\.\}\) = 0
close\(3\)                                = 0
open\("/lib/libc\.so\.6"\, O\_RDONLY\)        = 3
read\(3\, "\\177ELF\\1\\1\\1\\0\\0\\0\\0\\0\\0\\0\\0\\0\\3\\0\\3\\0\\1\\0\\0\\0\\0\\302\\1"\.\.\.\, 1024\) = 1024
fstat64\(3\, \{st\_mode=S\_IFREG|0755\, st\_size=1216268\, \.\.\.\}\) = 0
close\(3\)                                = 0
open\("/lib/libc\.so\.6"\, O\_RDONLY\)        = 3
read\(3\, "\\177ELF\\1\\1\\1\\0\\0\\0\\0\\0\\0\\0\\0\\0\\3\\0\\3\\0\\1\\0\\0\\0\\0\\302\\1"\.\.\.\, 1024\) = 1024
fstat64\(3\, \{st\_mode=S\_IFREG|0755\, st\_size=1216268\, \.\.\.\}\) = 0
close\(3\)                                = 0
open\("/lib/libc\.so\.6"\, O\_RDONLY\)        = 3
read\(3\, "\\177ELF\\1\\1\\1\\0\\0\\0\\0\\0\\0\\0\\0\\0\\3\\0\\3\\0\\1\\0\\0\\0\\0\\302\\1"\.\.\.\, 1024\) = 1024
fstat64\(3\, \{st\_mode=S\_IFREG|0755\, st\_size=1216268\, \.\.\.\}\) = 0
close\(3\)                                = 0
open\("/lib/libc\.so\.6"\, O\_RDONLY\)        = 3
read\(3\, "\\177ELF\\1\\1\\1\\0\\0\\0\\0\\0\\0\\0\\0\\0\\3\\0\\3\\0\\1\\0\\0\\0\\0\\302\\1"\.\.\.\, 1024\) = 1024
fstat64\(3\, \{st\_mode=S\_IFREG|0755\, st\_size=1216268\, \.\.\.\}\) = 0
close\(3\)                                = 0
open\("/lib/libc\.so\.6"\, O\_RDONLY\)        = 3
read\(3\, "\\177ELF\\1\\1\\1\\0\\0\\0\\0\\0\\0\\0\\0\\0\\3\\0\\3\\0\\1\\0\\0\\0\\0\\302\\1"\.\.\.\, 1024\) = 1024
fstat64\(3\, \{st\_mode=S\_IFREG|0755\, st\_size=1216268\, \.\.\.\}\) = 0
close\(3\)                                = 0
old\_mmap\(NULL\, 4096\, PROT\_READ|PROT\_WRITE\, MAP\_PRIVATE|MAP\_ANONYMOUS\, \-1\, 0\) = 0x401d2000
munmap\(0x40016000\, 43780\)               = 0
getpid\(\)                                = 11379
rt\_sigaction\(SIGRT\_0\, \{0x40066ad0\, \[\]\, 0x4000000\}\, NULL\, 8\) = 0
rt\_sigaction\(SIGRT\_1\, \{0x40065e80\, \[\]\, 0x4000000\}\, NULL\, 8\) = 0
rt\_sigaction\(SIGRT\_2\, \{0x40066b60\, \[\]\, 0x4000000\}\, NULL\, 8\) = 0
rt\_sigprocmask\(SIG\_BLOCK\, \[RT\_0\]\, NULL\, 8\) = 0
\_sysctl\(\{\{CTL\_KERN\, KERN\_VERSION\}\, 2\, 0xbffff51c\, 31\, \(nil\)\, 0\}\) = 0
getpid\(\)                                = 11379
rt\_sigaction\(SIGFPE\, \{SIG\_IGN\}\, \{SIG\_DFL\}\, 8\) = 0
brk\(0\)                                  = 0x8137740
brk\(0x8137770\)                          = 0x8137770
brk\(0x8138000\)                          = 0x8138000
brk\(0x8139000\)                          = 0x8139000
brk\(0x813a000\)                          = 0x813a000
brk\(0x813c000\)                          = 0x813c000
brk\(0x813d000\)                          = 0x813d000
SYS\_199\(0x4019bf2c\, 0\, 0x4019cb60\, 0x4019a9d0\, 0\) = 501
semop\(1075429164\, 0x4019a9d0\, 0\)        = 501
ipc\_subcall\(0x4019bf2c\, 0\, 0x4019cb60\, 0x4019a9d0\) = 501
semget\(1075429164\, 0\, IPC\_CREAT|IPC\_NOWAIT|0x4019c160|0440\) = 501
open\("/usr/share/locale/locale\.alias"\, O\_RDONLY\) = 3
fstat64\(3\, \{st\_mode=S\_IFREG|0644\, st\_size=2567\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 4096\, PROT\_READ|PROT\_WRITE\, MAP\_PRIVATE|MAP\_ANONYMOUS\, \-1\, 0\) = 0x40016000
read\(3\, "\# Locale name alias data base\.\\n\#"\.\.\.\, 4096\) = 2567
read\(3\, ""\, 4096\)                       = 0
close\(3\)                                = 0
munmap\(0x40016000\, 4096\)                = 0
open\("/usr/share/locale/en/LC\_IDENTIFICATION"\, O\_RDONLY\) = 3
fstat64\(3\, \{st\_mode=S\_IFREG|0644\, st\_size=244\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 244\, PROT\_READ\, MAP\_PRIVATE\, 3\, 0\) = 0x40016000
close\(3\)                                = 0
open\("/usr/share/locale/en/LC\_MEASUREMENT"\, O\_RDONLY\) = 3
fstat64\(3\, \{st\_mode=S\_IFREG|0644\, st\_size=13\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 13\, PROT\_READ\, MAP\_PRIVATE\, 3\, 0\) = 0x40017000
close\(3\)                                = 0
open\("/usr/share/locale/en/LC\_TELEPHONE"\, O\_RDONLY\) = 3
fstat64\(3\, \{st\_mode=S\_IFREG|0644\, st\_size=49\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 49\, PROT\_READ\, MAP\_PRIVATE\, 3\, 0\) = 0x40018000
close\(3\)                                = 0
open\("/usr/share/locale/en/LC\_ADDRESS"\, O\_RDONLY\) = 3
fstat64\(3\, \{st\_mode=S\_IFREG|0644\, st\_size=145\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 145\, PROT\_READ\, MAP\_PRIVATE\, 3\, 0\) = 0x40019000
close\(3\)                                = 0
brk\(0x813e000\)                          = 0x813e000
open\("/usr/share/locale/en/LC\_NAME"\, O\_RDONLY\) = 3
fstat64\(3\, \{st\_mode=S\_IFREG|0644\, st\_size=67\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 67\, PROT\_READ\, MAP\_PRIVATE\, 3\, 0\) = 0x4001a000
close\(3\)                                = 0
open\("/usr/share/locale/en/LC\_PAPER"\, O\_RDONLY\) = 3
fstat64\(3\, \{st\_mode=S\_IFREG|0644\, st\_size=24\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 24\, PROT\_READ\, MAP\_PRIVATE\, 3\, 0\) = 0x4001b000
close\(3\)                                = 0
open\("/usr/share/locale/en\_US/LC\_MESSAGES"\, O\_RDONLY\) = 3
fstat64\(3\, \{st\_mode=S\_IFDIR|0755\, st\_size=4096\, \.\.\.\}\) = 0
close\(3\)                                = 0
open\("/usr/share/locale/en\_US/LC\_MESSAGES/SYS\_LC\_MESSAGES"\, O\_RDONLY\) = 3
fstat64\(3\, \{st\_mode=S\_IFREG|0644\, st\_size=42\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 42\, PROT\_READ\, MAP\_PRIVATE\, 3\, 0\) = 0x4001c000
close\(3\)                                = 0
open\("/usr/share/locale/en\_US/LC\_MONETARY"\, O\_RDONLY\) = 3
fstat64\(3\, \{st\_mode=S\_IFREG|0644\, st\_size=276\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 276\, PROT\_READ\, MAP\_PRIVATE\, 3\, 0\) = 0x4001d000
close\(3\)                                = 0
open\("/usr/share/locale/en\_US/LC\_COLLATE"\, O\_RDONLY\) = 3
fstat64\(3\, \{st\_mode=S\_IFREG|0644\, st\_size=21484\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 21484\, PROT\_READ\, MAP\_PRIVATE\, 3\, 0\) = 0x401d3000
close\(3\)                                = 0
open\("/usr/share/locale/en\_US/LC\_TIME"\, O\_RDONLY\) = 3
fstat64\(3\, \{st\_mode=S\_IFREG|0644\, st\_size=2441\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 2441\, PROT\_READ\, MAP\_PRIVATE\, 3\, 0\) = 0x4001e000
close\(3\)                                = 0
open\("/usr/share/locale/en\_US/LC\_NUMERIC"\, O\_RDONLY\) = 3
fstat64\(3\, \{st\_mode=S\_IFREG|0644\, st\_size=44\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 44\, PROT\_READ\, MAP\_PRIVATE\, 3\, 0\) = 0x4001f000
close\(3\)                                = 0
open\("/usr/share/locale/en\_US/LC\_CTYPE"\, O\_RDONLY\) = 3
fstat64\(3\, \{st\_mode=S\_IFREG|0644\, st\_size=110304\, \.\.\.\}\) = 0
old\_mmap\(NULL\, 110304\, PROT\_READ\, MAP\_PRIVATE\, 3\, 0\) = 0x401d9000
close\(3\)                                = 0
brk\(0x8140000\)                          = 0x8140000
time\(\[1021638356\]\)                      = 1021638356
brk\(0x8141000\)                          = 0x8141000
open\("/dev/null"\, O\_RDONLY|O\_LARGEFILE\) = 3
shmat\(3\, 0x2\, 0x2ptrace&#8203;: umoven&#8203;: Input/output error
\)                      = ?
rt\_sigaction\(SIGCHLD\, NULL\, \{SIG\_DFL\}\, 8\) = 0
brk\(0x8142000\)                          = 0x8142000
brk\(0x8143000\)                          = 0x8143000
brk\(0x8144000\)                          = 0x8144000
readlink\("/proc/self/exe"\, "/usr/bin/perl5\.7\.3"\, 4094\) = 18
brk\(0x8145000\)                          = 0x8145000
brk\(0x8146000\)                          = 0x8146000
getpid\(\)                                = 11379
brk\(0x8147000\)                          = 0x8147000
brk\(0x8148000\)                          = 0x8148000
close\(3\)                                = 0
\-\-\- SIGSEGV \(Segmentation fault\) \-\-\-
\+\+\+ killed by SIGSEGV \+\+\+
~$ 
\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.1:

Configured by sbeck at Sat Mar  2 03:06:09 PST 2002.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.4.5, archname=i686-linux
    uname='linux athena 2.4.5 #2 fri mar 1 17:53:43 pst 2002 i686 unknown '
    config_args='-des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.3 20010315 (release)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.2.3.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.1:
    /usr/local/lib/perl5/5.6.1/i686-linux
    /usr/local/lib/perl5/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.1/i686-linux
    /usr/local/lib/perl5/site_perl/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.0/i686-linux
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.1:
    HOME=/home/sbeck
    LANG (unset)
    LANGUAGE (unset)
    LC_ALL=POSIX
    LD_LIBRARY_PATH=/usr/local/lib:/usr/lib:/usr/local/lib:/usr/lib:
    LOGDIR (unset)
    PATH=/home/sbeck/bin:/usr/local/bin:/opt/gnome/bin:/usr/local/mozilla:/usr/local/bin:/usr/bin:/bin:/usr/X11R6/bin:/usr/games:.:/opt/gnome/bin
    PERLDOC_PAGER=sh -c 'cat $0 | view -c "set ft=perldoc nomod nolist" -'
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @rgarcia

Scott Beck wrote&#8203;:
> This bug started appearing in perl5\.6\.1\. Perl segfaults when you delete a named
> subroutine from the symbol table\, and then call it with a reference to it\. Here
> is an example with the latest perl I have&#8203;:
> 
>     perl5\.7\.3 \-e 'sub foo \{my @&#8203;f = caller\(0\)\} my $code = delete $&#8203;::\{foo\}; $code\->\(\);'

I think that an acceptable fix would be to put the string "\(unknown\)"
\(or "\(deleted\)"\) in the subroutine\-name slot of the caller\(\) return
value\.

Where do the regression tests for caller\(\) go ?

\-\- 
Rafael Garcia\-Suarez

**Comment by p5pRT** [comment]
### From @jhi

On Fri\, May 17\, 2002 at 03&#8203;:53&#8203;:38PM \+0200\, Rafael Garcia\-Suarez wrote&#8203;:
> Scott Beck wrote&#8203;:
> > This bug started appearing in perl5\.6\.1\. Perl segfaults when you delete a named
> > subroutine from the symbol table\, and then call it with a reference to it\. Here
> > is an example with the latest perl I have&#8203;:
> > 
> >     perl5\.7\.3 \-e 'sub foo \{my @&#8203;f = caller\(0\)\} my $code = delete $&#8203;::\{foo\}; $code\->\(\);'
> 
> I think that an acceptable fix would be to put the string "\(unknown\)"
> \(or "\(deleted\)"\) in the subroutine\-name slot of the caller\(\) return
> value\.
> 
> Where do the regression tests for caller\(\) go ?

Time for t/op/caller\.t?

> \-\- 
> Rafael Garcia\-Suarez

\-\- 
$jhi\+\+; \# http&#8203;://www\.iki\.fi/jhi/
&nbsp;        \# There is this special biologist word we use for 'stable'\.
&nbsp;        \# It is 'dead'\. \-\- Jack Cohen

**Comment by p5pRT** [comment]
### From @rgarcia

On 2002\.05\.17 16&#8203;:00 Jarkko Hietaniemi wrote&#8203;:
> On Fri\, May 17\, 2002 at 03&#8203;:53&#8203;:38PM \+0200\, Rafael Garcia\-Suarez wrote&#8203;:
> > Scott Beck wrote&#8203;:
> > > This bug started appearing in perl5\.6\.1\. Perl segfaults when you delete a named
> > > subroutine from the symbol table\, and then call it with a reference to it\. Here
> > > is an example with the latest perl I have&#8203;:
> > > 
> > >     perl5\.7\.3 \-e 'sub foo \{my @&#8203;f = caller\(0\)\} my $code = delete $&#8203;::\{foo\}; $code\->\(\);'
> > 
> > I think that an acceptable fix would be to put the string "\(unknown\)"
> > \(or "\(deleted\)"\) in the subroutine\-name slot of the caller\(\) return
> > value\.
> > 
> > Where do the regression tests for caller\(\) go ?
> 
> Time for t/op/caller\.t?

This bug fixed by \#16658\, ant t/op/caller\.t added\.
Now caller\(\) returns '\(unknown\)' in this case\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

In article \<20020517220756\.A711@&#8203;ra\_ael>\,
&nbsp;    Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr> writes&#8203;:
> 
> This bug fixed by \#16658\, ant t/op/caller\.t added\.
> Now caller\(\) returns '\(unknown\)' in this case\.

Wouldn't undef be better ? "\(unknown\)" can in principle
be a valid function name if you play directly with
the globs\.

**Comment by p5pRT** [comment]
### From @rgarcia

On 2002\.05\.20 16&#8203;:33 Ton Hospel wrote&#8203;:
> In article \<20020517220756\.A711@&#8203;ra\_ael>\,
>     Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr> writes&#8203;:
> > 
> > This bug fixed by \#16658\, ant t/op/caller\.t added\.
> > Now caller\(\) returns '\(unknown\)' in this case\.
> 
> Wouldn't undef be better ? "\(unknown\)" can in principle
> be a valid function name if you play directly with
> the globs\.

Yes\, but in this case\, caller\(\) would return the fully
qualified name of the subroutine \("foo&#8203;::bar&#8203;::\(unknown\)"\)\.
So you can make the difference\.

I proposed "\(unknown\)" because caller\(\) may also return
"\(eval\)" :

&nbsp;    $ perl \-le 'eval\{print\(\(caller\(0\)\)\[3\]\)\}'
&nbsp;    \(eval\)

Moreover this information returned by caller\(\) is mainly
useful for debugging purposes : a plain string is more useful
than an undef\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'
